src/hotspot/share/runtime/thread.cpp

```diff
@@ -1585,157 +1581,109 @@ void JavaThread::remove_monitor_chunk(MonitorChunk* chunk) {
+      if (has_last_Java_frame()) {
+        frame f = last_frame();
+        ls.print(" (pc: " INTPTR_FORMAT " sp: " INTPTR_FORMAT " )", p2i(f.pc()), p2i(f.sp()));
       }
+      ls.print_cr(" of type: %s", java_throwable->klass()->external_name());
     }
-    // Always null out the _pending_async_exception oop here since the async condition was
-    // already cleared above and thus considered handled.
-    _pending_async_exception = NULL;
-  } else {
-    assert(_is_unsafe_access_error, "must be");
-    DEBUG_ONLY(_is_unsafe_access_error = false);
-
-    // We may be at method entry which requires we save the do-not-unlock flag.
-    UnlockFlagSaver fs(this);
-    Exceptions::throw_unsafe_access_internal_error(this, __FILE__, __LINE__, "a fault occurred in an unsafe memory access operation");
-    // We might have blocked in a ThreadBlockInVM wrapper in the call above so make sure we process pending
-    // suspend requests and object reallocation operations if any since we might be going to Java after this.
-    SafepointMechanism::process_if_requested_with_exit_check(this, true /* check asyncs */);
   }
 }
 
-void JavaThread::handle_special_runtime_exit_condition(bool check_asyncs) {
+void JavaThread::install_async_exception(AsyncExceptionHandshake* aeh) {
+  // Do not throw asynchronous exceptions against the compiler thread.
+  if (!can_call_java()) {
+    delete aeh;
+    return;
+  }
 
-  if (is_obj_deopt_suspend()) {
-    frame_anchor()->make_walkable(this);
-    wait_for_object_deoptimization();
+  // Don't install a new pending async exception if there is already
+  // a pending ThreadDeath one. Just interrupt thread from potential
+  // wait()/sleep()/park() and return.
+  if (has_async_exception_condition(true /* ThreadDeath_only */)) {
+    java_lang_Thread::set_interrupted(threadObj(), true);
+    this->interrupt();
+    delete aeh;
+    return;
   }
 
-  // We might be here for reasons in addition to the self-suspend request
-  // so check for other async requests.
-  if (check_asyncs) {
-    check_and_handle_async_exceptions();
+  oop exception = aeh->exception();
+  Handshake::execute(aeh, this);  // Install asynchronous handshake
+
+  ResourceMark rm;
+  if (log_is_enabled(Info, exceptions)) {
+    log_info(exceptions)("Pending Async. exception installed of type: %s",
+                         InstanceKlass::cast(exception->klass())->external_name());
   }
+  // for AbortVMOnException flag
+  Exceptions::debug_check_abort(exception->klass()->external_name());
 
-  JFR_ONLY(SUSPEND_THREAD_CONDITIONAL(this);)
+  // Interrupt thread so it will wake up from a potential wait()/sleep()/park()
+  java_lang_Thread::set_interrupted(threadObj(), true);
+  this->interrupt();
 }
 
-class InstallAsyncExceptionClosure : public HandshakeClosure {
-  Handle _throwable; // The Throwable thrown at the target Thread
+class InstallAsyncExceptionHandshake : public HandshakeClosure {
```

Comment:
Should this class still have Closure in its name?

I'm unclear why we need this class to wrap an AsyncExceptionHandShake ??

Reply:
I looked at the suspend handshakes (ThreadSelfSuspensionHandshake, SuspendThreadHandshake) and followed that naming convention. But now doing a grep I see that most handshakes have Closure instead at the end, some have HandshakeClosure, and some just have handshake at the beginning(HandshakeForDeflation, HandshakeForPD) . Not sure now which one we should use. It seems it should be either Closure or HandshakeClosure at the end.
Regarding InstallAsyncExceptionHandshake, we need it to execute the first handshake that installs the actual asynchronous handshake. The closure of that asynchronous handshake is stored in InstallAsyncExceptionHandshake::_aeh.

Reply:
Hmmm ... I'm having trouble following the control flow here so will re-examine.

The naming issue is a concern ... perhaps @robehn can comment?

Reply:
The name comes from that it extends ThreadClosure.
The reason for that is that GC have thread closures to be executed as handshakes or for each thread in a safepoint.

Whatever the name is good or bad I have no opinion.

Reply:
Ok, as Dan pointed out we can normalize all the names in a later bug. If we already decided on what should be the name I can change these two now.

Reply:
Lets normalise later please.

Reply:
Sounds good. Filed 8284016 to address that.